core/modules/block/block.module

```diff
@@ -273,22 +321,64 @@ function block_custom_block_form($edit = array(), $stand_alone = TRUE) {
  *   Machine name of the block to save, comprised of lower-case letters,
  *   numbers, and underscores. Note: NULL values are still accepted when
  *   creating new nodes to maintain API contiguity.
+ * @param $langcode
+ *   The language code for the block.
+ *   If set, only the translation is saved.
+ *
+ * Custom blocks have two new keys in the config.
+ * - default_langcode and
+ * - translations
+ *
+ * default_langcode holds the langcode of the source language - the language the block is created with.
+ * In the translations key the translations of the other languages are stored.
+ * The structure of the tranlations is:
+ * [langcode] => array(
+ *    'info' => '',
+ *    'title' => '',
+ *    'body' => array('value' => '', 'format' => ''),
+ * )
  */
-function block_custom_block_save(array $edit, $delta = NULL) {
+function block_custom_block_save(array $edit, $delta = NULL, $langcode = NULL) {
+  $translateable = module_exists('locale');
+
   $delta = $delta ? $delta : preg_replace('/[^a-z0-9_]+/', '_', strtolower($edit['info']));
   $block = block_custom_block_load($delta);
   if (!$block) {
     $block = array(
       'delta' => $delta,
       'info' => '',
       'title' => '',
-      'description' => '',
       'body' => array('value' => '', 'format' => ''),
+      'default_langcode' => LANGUAGE_NONE,
     );
   }
-  foreach ($block as $key => $value) {
-    if (isset($edit[$key])) {
-      $block[$key] = $edit[$key];
+  // If langcode == NULL we come from add or edit form, so
+  // store the default values.
+  // Same if the langcode is the default language.
+  if ($langcode === NULL || (isset($block['default_langcode']) &&
+      ($langcode == $block['default_langcode']))) {
+    foreach ($block as $key => $value) {
+      if (isset($edit[$key])) {
+        $block[$key] = $edit[$key];
+      }
+    }
+  }
+
+  // if langcode != NULL we come from a translation form, so
+  // store the translation.
+  if ($translateable &&
+      ($langcode != NULL) &&
+      isset($block['default_langcode']) &&
+      ($langcode != $block['default_langcode'])) {
+    // store the translations
```

Comment:
Minor code style things, code comments should be Sentence case and end with a period.